### PR TITLE
Fun with UINT64_C

### PR DESCRIPTION
**Pull request by keszybz**
Sometimes C just makes me want to scratch my head.

### CODE REVIEW

File: src/nspawn/nspawn-settings.h [hunk 1]
```diff
@@ -56,9 +56,19 @@ typedef enum SettingsMask {
...
         SETTING_RLIMIT_FIRST      = UINT64_C(1) << 21, /* we define one bit per resource limit here */
         SETTING_RLIMIT_LAST       = UINT64_C(1) << (21 + _RLIMIT_MAX - 1),
-        _SETTINGS_MASK_ALL        = (UINT64_C(1) << (21 + _RLIMIT_MAX)) - 1
+        _SETTINGS_MASK_ALL        = (UINT64_C(1) << (21 + _RLIMIT_MAX)) - 1,
+        _FORCE_ENUM_WIDTH         = UINT64_C(-1)
 } SettingsMask;
 
+/* We want to use SETTING_RLIMIT_FIRST in shifts, so make sure it is really 64 bits
+ * when used in expressions. */
+#define SETTING_RLIMIT_FIRST (UINT64_C(1) << 21)
```
Comment (poettering) [inline src/nspawn/nspawn-settings.h]: maybe define as: 

```
#define SETTING_RLIMIT_FIRST ((uint64_t) SETTING_RLIMIT_FIRST)
#define SETTING_RLIMIT_LAST ((uint64_t) SETTING_RLIMIT_LAST)
``` 

that way we would get rid of the double definition of the value, and whenever we add new flags we don't have to update these two places too

File: src/nspawn/nspawn-settings.h [hunk 2]
```diff
@@ -56,9 +56,19 @@ typedef enum SettingsMask {
         SETTING_CPU_AFFINITY      = UINT64_C(1) << 20,
         SETTING_RLIMIT_FIRST      = UINT64_C(1) << 21, /* we define one bit per resource limit here */
         SETTING_RLIMIT_LAST       = UINT64_C(1) << (21 + _RLIMIT_MAX - 1),
-        _SETTINGS_MASK_ALL        = (UINT64_C(1) << (21 + _RLIMIT_MAX)) - 1
+        _SETTINGS_MASK_ALL        = (UINT64_C(1) << (21 + _RLIMIT_MAX)) - 1,
+        _FORCE_ENUM_WIDTH         = UINT64_C(-1)
```
Comment (poettering) [inline src/nspawn/nspawn-settings.h]: hmm, wouldn't it be nicer to define this as UINT64_MAX instead? I think that would be a lot more readable as defining a literal negative value as unsigned always strikes me very weird...

File: src/nspawn/nspawn-settings.h (hunk 1 above)
Comment (keszybz) [reply src/nspawn/nspawn-settings.h]: Hm, I was sure that wouldn't work because of a recursive definition, but it does. I'll update.

**Comment by keszybz** [comment]
Updated.

**Comment by yuwata** [comment]
lgtm.

**Comment by keszybz** [comment]
@yuwata thanks.

@evverx could you test that this work? I cannot reproduce the issue with ubsan... It doesn't give me an error that you showed.

File: src/nspawn/nspawn-settings.h [hunk 3]
```diff
@@ -56,9 +56,19 @@ typedef enum SettingsMask {
         SETTING_CPU_AFFINITY      = UINT64_C(1) << 20,
         SETTING_RLIMIT_FIRST      = UINT64_C(1) << 21, /* we define one bit per resource limit here */
         SETTING_RLIMIT_LAST       = UINT64_C(1) << (21 + _RLIMIT_MAX - 1),
-        _SETTINGS_MASK_ALL        = (UINT64_C(1) << (21 + _RLIMIT_MAX)) - 1
+        _SETTINGS_MASK_ALL        = (UINT64_C(1) << (21 + _RLIMIT_MAX)) - 1,
+        _FORCE_ENUM_WIDTH         = UINT64_MAX
```
Comment (poettering) [inline src/nspawn/nspawn-settings.h:60]: btw, i figure one of those days we should probably rename this to _SETTINGS_MASK_FORCE_ENUM_WIDTH or so, i.e. add the namespacing prefix, as we might want to do the same for other enums too and then the names would clash

**Comment by poettering** [comment]
lgtm

**Comment by filbranden** [comment]
Frankly... Are enums better than this?

```c
#define SETTING_START_MODE   (1L << 0)
#define SETTING_ENVIRONMENT  (1L << 1)
/* ... */
/* we define one bit per resource limit here */
#define SETTING_RLIMIT_FIRST (1L << 21)
#define SETTING_RLIMIT_LAST  (1L << (21 + _RLIMIT_MAX - 1))
typedef uint64_t SettingsMask;
```

The C enum doesn't really provide any type checking whatsoever and ends up just creating a very implicit limitation on the width of the fields...

On the other PR, I already raised the question of whether using enums for bitfields is really acceptable. (For one, C++ with enum classes doesn't allow it... They're different languages, particularly in this area, but it seems to me the point of enums is to set to one of the alternatives, not use bit operations on them.)

While the `_FORCE_ENUM_WIDTH` seems to work (with limitations, you still need the cast when using the enum values), is it really worth it?

Cheers,
Filipe

**Comment by filbranden** [comment]
Or, in other words... #9070.

**Comment by keszybz** [comment]
So enums have the important advantage the the debugger knows about them. That's why I kept the enum. But like you say, I'm not sure if it is worth the extra complexity.

**Pull request merged**